editor/sidebar/post-visibility/index.js

```diff
@@ -24,62 +24,74 @@ import { editPost, savePost } from '../../actions';
 class PostVisibility extends Component {
 	constructor( props ) {
 		super( ...arguments );
+
+		this.toggleDialog = this.toggleDialog.bind( this );
+		this.setVisibility = this.setVisibility.bind( this );
+
 		this.state = {
 			opened: false,
 			hasPassword: !! props.password,
 		};
-		this.toggleDialog = this.toggleDialog.bind( this );
 	}
 
 	toggleDialog( event ) {
 		event.preventDefault();
 		this.setState( { opened: ! this.state.opened } );
 	}
 
+	setVisibility( event ) {
+		const { visibility, onUpdateVisibility, status, password, onSave } = this.props;
+		const nextVisibility = event.currentTarget.value;
+
+		switch ( nextVisibility ) {
+			case 'public':
+				onUpdateVisibility( visibility === 'private' ? 'draft' : status );
+				this.setState( { hasPassword: false } );
+				break;
+
+			case 'private':
+				if ( ! window.confirm( __( 'Would you like to privately publish this post now?' ) ) ) { // eslint-disable-line no-alert
+					return;
+				}
+
+				onUpdateVisibility( 'private' );
+				onSave();
+				this.setState( { opened: false } );
+				break;
+
+			case 'password':
+				onUpdateVisibility( visibility === 'private' ? 'draft' : status, password || '' );
+				this.setState( { hasPassword: true } );
+				break;
+		}
+	}
```

Comment:
I don't have strong arguments but I'd have preferred three functions instead of a switch :)